File: Projects/Lightning/Source/Tracker/lightning.py
import struct, tracker_pb2, lightning_pb2

tracker_address = "95.80.1.56"
tracker_port = 50000

# Common enumerators
OK, PING, PONG, GOODBYE = range(0,4)

# Tracker <-> Client
T_INIT, T_HELLO_CLIENT, T_HELLO_TRACKER, T_NEWBATCH, T_NODES_REQUEST \
    , T_NODES_RESPONSE, T_BATCH_DONE, T_TASK_DONE, T_HATE = range(4, 13)

# Clinet <-> Client
C_HELLO, C_ALLOCATE, C_ALLOC_RESP, C_DATA, C_TASKS = range(13, 18)



translate_message_type = {
    T_INIT: None,
    T_HELLO_CLIENT: tracker_pb2.HelloToClient,
    T_HELLO_TRACKER: tracker_pb2.HelloToTracker,
    GOODBYE: lightning_pb2.GoodBye,
    PING: lightning_pb2.Ping,
    PONG: lightning_pb2.Pong
}

def unpack_raw(message):
    """unpack_raw(str message) -> (int message_type, string data)

    Wrapper for unpacking data sent between the tracker and client.
    """

    # The first 4 bytes of the message is an identifier
    type_data = struct.unpack('<I', message[:4])[0]
    return (type_data, message[4:])

def unpack(message):
    """unpack(str message) -> (int message_type, <protobuf> data)

    Returns a tuple containing the type of data, and parsed protocol buffers
    data.
    """

    (type_data, data) = unpack_raw(message)

    if type_data not in translate_message_type:
        raise PackingException("Enum {0} could not be found in translate dict".format(type_data))

    if data == "":
        return (type_data, None)
    else:
        if translate_message_type[type_data] is not None:
            proto_data = translate_message_type[type_data]()
        else:
            return (type_data, None)
        proto_data.ParseFromString(data)
        return (type_data, proto_data)


def pack(message_type, data=None):
    """pack(int message_type, <protobuf> data=None) -> str message.

    Packs a message type and data and returns a message string.
    If no data is to be sent, set it to NULL.
    """

    ret = struct.pack('<I', message_type)
    if data is not None:
        ret += data.SerializeToString()

    return ret

class LightningException(Exception):
    """Main class for lightnings exceptions, if there are any..."""
    pass

class PackingException(LightningException):
    pass

File: Projects/Lightning/Source/ClientLib/lightning.cpp
#include "lightning.h"

#include <stdio.h>

google::protobuf::Message *get_protobuf_class(MessageType type)
{
	switch(type)
	{
		case GOODBYE: return new lightning::GoodBye;
		case PING: return new lightning::Ping;
		case PONG: return new lightning::Pong;
		case OK: return(NULL);

		case T_INIT: return(NULL);
		case T_HELLO_CLIENT: return new tracker::HelloToClient;
		case T_HELLO_TRACKER: return new tracker::HelloToTracker;
		case T_NEWBATCH: return new tracker::NewBatch;
		case T_NODES_REQUEST: return new tracker::NodesRequest;
		case T_NODES_RESPONSE: return new tracker::NodesResponse;
		case T_BATCH_DONE: return NULL; // TODO: Fill in
		case T_TASK_DONE: return NULL;
		case T_HATE: return NULL;

		case C_HELLO: return new client::Hello;
		case C_ALLOCATE: return new client::AllocateClient;
		case C_ALLOC_RESP: return new client::AllocateResponse;
		case C_DATA: return new client::Data;
		case C_TASKS: return new client::Tasks;
		case C_RESULT: return new client::Result;
	}

	return NULL;
}

LiPacket::LiPacket(Connection *_c, google::protobuf::Message *_proto, int _type)
{
	connection = _c;
	pack(_proto, _type);

	__pkg = NULL;
}

LiPacket::LiPacket(Packet* _pkg)
{
	__pkg = _pkg;
	data = _pkg->data;
	length = _pkg->length;
	connection = _pkg->connection;
}

LiPacket::~LiPacket()
{
	if (__pkg)
		delete __pkg;	
}

MessageType LiPacket::get_type()
{
	message_type = (MessageType)((int*)data)[0];//(MessageType)*(int*)data;
	return message_type;
}


char *LiPacket::pack(google::protobuf::Message *_proto, int _type)
{
	length = sizeof(_type) + _proto->ByteSize();
	data = (char*)Pxf::MemoryAllocate(length);

	Pxf::MemoryCopy(
		data,
		&(_type),
		sizeof(_type)
	);

	Pxf::MemoryCopy(
		data+sizeof(_type),
		_proto->SerializeAsString().c_str(),
		_proto->ByteSize()
	);

	message_type = (MessageType)_type;

	return data;
}

google::protobuf::Message *LiPacket::unpack()
{
	Pxf::MemoryCopy(
		(MessageType*)&message_type,
		data,
		sizeof(message_type)
	);

	google::protobuf::Message *proto = get_protobuf_class(message_type);

	if (proto == NULL) return NULL;

	proto->ParseFromString(Pxf::Util::String(data+sizeof(message_type), length-sizeof(message_type)));

	return proto;
}

